### PR TITLE
refactor(cp): clean up code and fix #376

### LINKED ISSUE #376: Copy not accepting wildcards + file ending anymore when using -r
So while `ls("src/*.html")` returns an array of html files, `copy("-r", "src/*.html", "dist")` results in an error. It works without -r.
I'm on Windows 7 x64, Node 5.7, using 0.6.

### PR DESCRIPTION
**Pull request by nfischer**
Fixes issue in #376. Simplifies the code, and slight perf win.

This fixes the semantics of `cp()`, which had deviated from unix `cp` a bit. If anyone is available to review, that would be good.

Goals of this PR:

 - improve performance
 - improve correctness

### CODE REVIEW

**Comment by nfischer** [comment]
@ariporad can you take a look at this? Let me know what you think, or if there's any other perf improvements we can squeeze out.

File: test/set.js [hunk 1]
```diff
@@ -6,7 +6,6 @@ var oldConfigSilent = shell.config.silent;
 shell.config.silent = true;
 
 shell.rm('-rf', 'tmp');
-shell.mkdir('tmp');
```
Comment (ariporad) [inline test/set.js]: Why is this here? It seems completely unrelated.

**Comment by ariporad** [comment]
@nfischer: One tiny nit, and could you please rebase off `master`? Otherwise LGTM.

File: test/set.js (hunk 1 above)
Comment (nfischer) [inline test/set.js]: One of the tests assumed incorrect semantics of `cp()`. When I fixed the semantics, I realized this test was failing.

The directory actually gets created with a `cp()` command on [this line](https://github.com/shelljs/shelljs/blob/fix-cp-r-option/test/set.js#L20)

Comment (ariporad) [inline test/set.js]: Ok.

**Comment by ariporad** [comment]
@nfischer: Can you rebase of master? Otherwise LGTM!!!!

**Comment by nfischer** [comment]
@ariporad I just re-pushed
